File: libs/langchain-nia/langchain_nia/toolkit.py
"""Nia toolkit providing all Nia tools for LangChain agents."""

from __future__ import annotations

from typing import List

from langchain_core.tools import BaseTool, BaseToolkit
from pydantic import Field

from langchain_nia._api_wrapper import NiaAPIWrapper
from langchain_nia.advisor import NiaAdvisor
from langchain_nia.context_save import NiaContextSave
from langchain_nia.context_search import NiaContextSearch
from langchain_nia.deep_research import NiaDeepResearch
from langchain_nia.dependency_analyze import NiaDependencyAnalyze
from langchain_nia.dependency_subscribe import NiaDependencySubscribe
from langchain_nia.github_glob import NiaGitHubGlob
from langchain_nia.github_read import NiaGitHubRead
from langchain_nia.github_search import NiaGitHubSearch
from langchain_nia.github_tree import NiaGitHubTree
from langchain_nia.search import NiaSearch
from langchain_nia.source_explore import NiaExplore
from langchain_nia.source_grep import NiaGrep
from langchain_nia.source_index import NiaIndex
from langchain_nia.source_list import NiaSourceList
from langchain_nia.source_read import NiaRead
from langchain_nia.source_subscribe import NiaSourceSubscribe
from langchain_nia.source_sync import NiaSourceSync
from langchain_nia.universal_search import NiaUniversalSearch
from langchain_nia.web_search import NiaWebSearch


class NiaToolkit(BaseToolkit):
    """Toolkit providing all Nia AI tools for LangChain agents.

    Use ``include_*`` flags to control which tool groups are returned.
    All tools share a single :class:`NiaAPIWrapper` instance.
    """

    api_wrapper: NiaAPIWrapper = Field(default_factory=NiaAPIWrapper)
    include_search: bool = True
    include_sources: bool = True
    include_github: bool = True
    include_contexts: bool = True
    include_dependencies: bool = True

    def get_tools(self) -> List[BaseTool]:
        tools: List[BaseTool] = []
        w = self.api_wrapper

        if self.include_search:
            tools.extend(
                [
                    NiaSearch(api_wrapper=w),
                    NiaWebSearch(api_wrapper=w),
                    NiaDeepResearch(api_wrapper=w),
                    NiaUniversalSearch(api_wrapper=w),
                    NiaAdvisor(api_wrapper=w),
                ]
            )

        if self.include_sources:
            tools.extend(
                [
                    NiaIndex(api_wrapper=w),
                    NiaSourceList(api_wrapper=w),
                    NiaSourceSubscribe(api_wrapper=w),
                    NiaSourceSync(api_wrapper=w),
                    NiaRead(api_wrapper=w),
                    NiaGrep(api_wrapper=w),
                    NiaExplore(api_wrapper=w),
                ]
            )

        if self.include_github:
            tools.extend(
                [
                    NiaGitHubSearch(api_wrapper=w),
                    NiaGitHubRead(api_wrapper=w),
                    NiaGitHubGlob(api_wrapper=w),
                    NiaGitHubTree(api_wrapper=w),
                ]
            )

        if self.include_contexts:
            tools.extend(
                [
                    NiaContextSave(api_wrapper=w),
                    NiaContextSearch(api_wrapper=w),
                ]
            )

        if self.include_dependencies:
            tools.extend(
                [
                    NiaDependencySubscribe(api_wrapper=w),
                    NiaDependencyAnalyze(api_wrapper=w),
                ]
            )

        return tools

File: libs/langchain-nia-js/src/toolkit.ts
import { StructuredTool } from "@langchain/core/tools";
import { NiaAPIWrapper, type NiaAPIWrapperParams } from "./utils.js";
import {
  NiaSearch,
  NiaWebSearch,
  NiaDeepResearch,
  NiaUniversalSearch,
  NiaAdvisor,
  NiaIndex,
  NiaSourceList,
  NiaSourceSubscribe,
  NiaSourceSync,
  NiaRead,
  NiaGrep,
  NiaExplore,
  NiaGitHubSearch,
  NiaGitHubRead,
  NiaGitHubGlob,
  NiaGitHubTree,
  NiaContextSave,
  NiaContextSearch,
  NiaDependencySubscribe,
  NiaDependencyAnalyze,
} from "./tools.js";

export interface NiaToolkitParams extends NiaAPIWrapperParams {
  apiWrapper?: NiaAPIWrapper;
  includeSearch?: boolean;
  includeSources?: boolean;
  includeGithub?: boolean;
  includeContexts?: boolean;
  includeDependencies?: boolean;
}

export class NiaToolkit {
  apiWrapper: NiaAPIWrapper;

  includeSearch: boolean;

  includeSources: boolean;

  includeGithub: boolean;

  includeContexts: boolean;

  includeDependencies: boolean;

  constructor(params: NiaToolkitParams = {}) {
    this.apiWrapper =
      params.apiWrapper ?? new NiaAPIWrapper(params);
    this.includeSearch = params.includeSearch ?? true;
    this.includeSources = params.includeSources ?? true;
    this.includeGithub = params.includeGithub ?? true;
    this.includeContexts = params.includeContexts ?? true;
    this.includeDependencies = params.includeDependencies ?? true;
  }

  getTools(): StructuredTool[] {
    const tools: StructuredTool[] = [];
    const w = { apiWrapper: this.apiWrapper };

    if (this.includeSearch) {
      tools.push(
        new NiaSearch(w),
        new NiaWebSearch(w),
        new NiaDeepResearch(w),
        new NiaUniversalSearch(w),
        new NiaAdvisor(w)
      );
    }

    if (this.includeSources) {
      tools.push(
        new NiaIndex(w),
        new NiaSourceList(w),
        new NiaSourceSubscribe(w),
        new NiaSourceSync(w),
        new NiaRead(w),
        new NiaGrep(w),
        new NiaExplore(w)
      );
    }

    if (this.includeGithub) {
      tools.push(
        new NiaGitHubSearch(w),
        new NiaGitHubRead(w),
        new NiaGitHubGlob(w),
        new NiaGitHubTree(w)
      );
    }

    if (this.includeContexts) {
      tools.push(new NiaContextSave(w), new NiaContextSearch(w));
    }

    if (this.includeDependencies) {
      tools.push(
        new NiaDependencySubscribe(w),
        new NiaDependencyAnalyze(w)
      );
    }

    return tools;
  }
}
